Lens.idr

```diff
@@ -1,125 +1,85 @@
-
-pos : Store s a -> s
-pos (MkStore _ s) = s
-
-peek : s -> Store s a -> a
-peek s (MkStore f _) = f s
-
-peeks : (s -> s) -> Store s a -> a
-peeks f (MkStore g s) = g (f s)
-
--- Lenses
+_1 : Functor f => Lens (a, c) (b, c) a b {f}
+_1 = lens fst (\(_, y) => \x => (x, y))
 
-data Lens a b = MkLens (a -> Store b a)
+_2 : Functor f => Lens (a, b) (a, c) b c {f}
+_2 = lens snd (\(x, _) => \y => (x, y))
 
-instance Category Lens where
-  id = MkLens (MkStore id)
-  (.) (MkLens f) (MkLens g) = MkLens (\a => case g a of
-    MkStore ba b => case f b of
-      MkStore cb c => MkStore (Prelude.Basics.(.) ba cb) c)
+infixr 4 $~
 
-lens : (a -> b) -> (b -> a -> a) -> Lens a b
-lens f g = MkLens (\a => MkStore (\b => g b a) (f a))
+($~) : Lens s t a b -> (a -> b) -> s -> t
+($~) = id
 
-iso : (a -> b) -> (b -> a) -> Lens a b
-iso f g = MkLens (\a => MkStore g (f a))
+infixr 4 .~
 
-getL : Lens a b -> a -> b
-getL (MkLens f) a = pos (f a)
+(.~) : Lens s t a b -> b -> s -> t
+(.~) = (. const)
 
-setL : Lens a b -> b -> a -> a
-setL (MkLens f) b = peek b . f
+infixr 4 <$~
 
-modL : Lens a b -> (b -> b) -> a -> a
-modL (MkLens f) g = peeks g . f
+(<$~) : Lens s t a b {f = Pair b} -> (a -> b) -> s -> (b, t)
+(<$~) l f = l (\x => (f x, f x))
 
-mergeL : Lens a c -> Lens b c -> Lens (Either a b) c
-mergeL (MkLens f) (MkLens g) = MkLens $ either (\a => map Left $ f a)
-                                               (\b => map Right $ g b)
+infixr 4 <.~
 
-infixr 0 ^$
-(^$) : Lens a b -> a -> b
-(^$) = getL
+(<.~) : Lens s t a b {f = Pair b} -> b -> s -> (b, t)
+(<.~) l x = l <$~ const x
 
-infixr 4 ^=
-(^=) : Lens a b -> b -> a -> a
-(^=) = setL
+infix 4 $=
 
-infixr 4 ^%=
-(^%=) : Lens a b -> (b -> b) -> a -> a
-(^%=) = modL
+($=) : Lens s t a b -> (a -> b) -> { [STATE s] ==> [STATE t] } Eff ()
+l $= f = updateM (l $~ f)
 
-fstLens : Lens (a,b) a
-fstLens = MkLens $ \(a,b) => MkStore (\ a' => (a', b)) a
+infix 4 .=
+(.=) : Lens s t a b -> b -> { [STATE s] ==> [STATE t] } Eff ()
+l .= x = updateM (l .~ x)
 
-sndLens : Lens (a,b) b
-sndLens = MkLens $ \(a,b) => MkStore (\ b' => (a, b')) b
+infix 4 <$=
+(<$=) : Lens s t a b {f = Pair b} ->
+        (a -> b) ->
+        { [STATE s] ==> [STATE t] } Eff b
+l <$= f = case (l <$~ f) !get of
+    (r, s) => do putM s
+                 return r
 
--- Partial lenses
+infix 4 <.=
+(<.=) : Lens s t a b {f = Pair b} -> b -> { [STATE s] ==> [STATE t] } Eff b
+l <.= x = l <$= const x
 
-data PLens a b = MkPLens (a -> Maybe (Store b a))
-
-instance Category PLens where
-  id = MkPLens (Just . MkStore id)
-  (.) (MkPLens f) (MkPLens g) = MkPLens (\a => do
-    MkStore wba b <- g a
-    MkStore wcb c <- f b
-    return (MkStore (wba . wcb) c))
-
-plens : (a -> Either a (Store b a)) -> PLens a b
-plens f = MkPLens $ either (const Nothing) Just . f
-
-getPL : PLens a b -> a -> Maybe b
-getPL (MkPLens f) a = map pos (f a)
-
-justPL : PLens (Maybe a) a
-justPL = MkPLens (\ma => do
-  a <- ma
-  return (MkStore Just a))
+private
```

Comment:
Idris has confusing rules around this. Since you've marked one identifier as private you have to mark the ones which are meant to be public:

https://github.com/idris-lang/Idris-dev/issues/1879

Sadly, it's all or nothing.